feat: Create versioning for segments

zachaysan opened the pull request:
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

This introduces versioning for segments by cloning their features into new objects and filtering out older versions of segment through a new manager class.

## How did you test this code?

A few new unit tests around the models as well as updates to some existing tests to handle audit log cases.

matthewelwell commented on api/segments/serializers.py:
```diff
@@ -81,11 +81,15 @@ def create(self, validated_data):
         metadata_data = validated_data.pop("metadata", [])
         self.validate_segment_rules_conditions_limit(rules_data)
+
+        # Create a version of the segment now that we're updating.
+        instance.deep_clone()
```
We might need to be more clever here... I'm worried about failure modes in the update that would mean we end up with phantom versions. 

We could wrap the below in a `try` / `except` perhaps, or we could move this logic to the model.

zachaysan replied:
Do you mean a `try` / `except` in the logic that handles updates? If so, we'd be altering the segment before cloning it which would defeat the purpose, no?

matthewelwell replied:
Well, yes, I guess in an ideal world, we'd create the clone without persisting it to the database before doing the update and then persist it once the update has happened successfully. 

Since that's not really possible, due to all the foreign key relationships, we could instead just handle the deletion (I guess we'd want to hard delete too) of the already created clone in the `except` - although this still isn't bulletproof.

zachaysan replied:
Ok I've implemented this. Sorta feels a bit cludgy but I guess it's the only way forward.

zachaysan commented on api/segments/models.py:
```diff
@@ -140,6 +207,51 @@ def get_segment(self):
 
+    def deep_clone(self, versioned_segment: Segment) -> "SegmentRule":
+        if self.rule:
+            assert False, "Unexpected rule, expecting segment set not rule"
```
Ok I've added a comment to explain it.

matthewelwell replied:
Hmm... I'd argue that the comment doesn't explain it though, it just states the same thing that one can interpret from the code. What I'm looking for is the 'why' - why are we only expecting a rule here attached to a segment, not a rule.

zachaysan replied:
My understanding is that a rule either belongs to a segment xor it belongs to a rule. There are no exceptions. I've made it more explicit in the comment.

zachaysan commented on api/segments/models.py:
```diff
@@ -84,6 +111,46 @@ def id_exists_in_rules_data(rules_data: typing.List[dict]) -> bool:
+        self.save()
+
+    def deep_clone(self) -> "Segment":
+        new_segment = Segment.objects.create(
+            name=self.name,
+            description=self.description,
+            project=self.project,
+            feature=self.feature,
+            version=self.version,
+            version_of=self,
+        )
```
Ok I'll implement that strategy. Just to double check, the timestamps will still be set to `now` for the new model, right?

zachaysan replied:
Ok this is done.

matthewelwell replied:
>Just to double check, the timestamps will still be set to now for the new model, right?

Good question - I don't know. We should verify in a test. Although I guess the question is whether we want that to be the case. Since we're backfilling the history, wouldn't we _want_ e.g. the `created_at` not to change?

zachaysan replied:
Well it turns out that none of the models involved in this PR have `created_at` or `updated_at` timestamps, although they probably should, at least on the segments. Should I add them to this PR? If so, what should the default be for existing objects `None` or `now()`? Should I include conditions and rules in having timestamps as well?

matthewelwell replied:
Let's add these fields to all models in the segments app, yes, but let's do it in a separate PR. Regarding the default, we _could_ be smart and look in the objects history but that might end up being a fairly mammoth migration.

zachaysan replied:
Ok sounds good I've made a new ticket for this.

https://github.com/Flagsmith/flagsmith/issues/4211

zachaysan replied:
The PR is up for the timestamps:

https://github.com/Flagsmith/flagsmith/pull/4236

zachaysan commented on api/segments/models.py:
```diff
@@ -84,6 +111,46 @@ def id_exists_in_rules_data(rules_data: typing.List[dict]) -> bool:
+        itself including itself.
+        """
+        self.version_of = self
+        self.save()
```
I tried setting the model to itself before create and it threw an exception. I think our only option is to keep this behaviour.

matthewelwell commented on api/tests/integration/audit/test_audit_logs.py:
```diff
@@ -250,7 +250,7 @@ def test_retrieve_audit_log_includes_changes_when_segment_override_created_and_d
     assert get_audit_logs_response_2.status_code == status.HTTP_200_OK
     results = get_audit_logs_response_2.json()["results"]
-    assert len(results) == 5
+    assert len(results) == 6
```
What's the audit log record that we've added here?

zachaysan replied:
I've fixed this back to the original value since our conversation about the versioned segments impact on the audit log

matthewelwell commented on api/tests/unit/segments/test_unit_segments_models.py:
```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
+        feature=feature,
+    )
+
+    # Check that the versioning is correct, since we'll testing
```
```suggestion
    # Check that the versioning is correct, since we'll be testing
```

matthewelwell commented on api/tests/unit/segments/test_unit_segments_models.py:
```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
+    assert segment.version == 1
+    assert segment.version_of == segment
+
+    parent_rule1 = SegmentRule.objects.create(
+        segment=segment, type=SegmentRule.ALL_RULE
+    )
+    parent_rule2 = SegmentRule.objects.create(
+        segment=segment, type=SegmentRule.ANY_RULE
+    )
+
+    child_rule1 = SegmentRule.objects.create(
+        rule=parent_rule1, type=SegmentRule.ANY_RULE
+    )
+    child_rule2 = SegmentRule.objects.create(
+        rule=parent_rule1, type=SegmentRule.ALL_RULE
+    )
+    child_rule3 = SegmentRule.objects.create(
+        rule=parent_rule2, type=SegmentRule.NONE_RULE
+    )
+    child_rule4 = SegmentRule.objects.create(
+        rule=parent_rule2, type=SegmentRule.ALL_RULE
+    )
```
I think we can stick to the structure created by the dashboard here which is just: 

```
parent_rule = SegmentRule.objects.create(
    segment=segment, type=SegmentRule.ALL_RULE,
)

child_rule1 = SegmentRule.objects.create(
    rule=parent_rule, type=SegmentRule.ANY_RULE
)
child_rule2 = SegmentRule.objects.create(
    rule=parent_rule, type=SegmentRule.NONE_RULE
)
...
```

Note that the UI only allows 'any' and 'none' rules nested beneath a single 'all' rule.

zachaysan replied:
Sounds good. Fixed locally.

matthewelwell commented on api/tests/unit/segments/test_unit_segments_models.py:
```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
+        rule=parent_rule2, type=SegmentRule.ALL_RULE
+    )
+
+    condition1 = Condition.objects.create(
+        rule=parent_rule1,
+        property="parent_rule1",
+        operator=EQUAL,
+        value="condition1",
+        created_with_segment=True,
+    )
+    condition2 = Condition.objects.create(
+        rule=parent_rule1,
+        property="parent_rule1",
+        operator=PERCENTAGE_SPLIT,
+        value="0.1",
+        created_with_segment=False,
+    )
```
Again, the UI doesn't allow for conditions at the parent rule level.

matthewelwell commented on api/tests/unit/segments/test_unit_segments_models.py:
```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
+        rule=parent_rule2, type=SegmentRule.ALL_RULE
+    )
+
+    condition1 = Condition.objects.create(
+        rule=parent_rule1,
+        property="parent_rule1",
+        operator=EQUAL,
+        value="condition1",
+        created_with_segment=True,
+    )
+    condition2 = Condition.objects.create(
+        rule=parent_rule1,
+        property="parent_rule1",
+        operator=PERCENTAGE_SPLIT,
+        value="0.1",
+        created_with_segment=False,
+    )
```
Again, the UI doesn't allow for conditions at the parent rule level.

zachaysan replied:
Sounds good. Fixed this locally.

matthewelwell commented on api/tests/unit/segments/test_unit_segments_models.py:
```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
+    )
+
+    # When
+    new_segment = segment.deep_clone()
```
I think we could perhaps improve the naming here and use `cloned_segment` instead of `new_segment` ?

matthewelwell commented on api/tests/unit/segments/test_unit_segments_models.py:
```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
+    assert segment.version == 2
+    assert segment.version_of == segment
+
+    new_segment = segment.deep_clone()
```
As above, let's use `cloned_segment` here?

matthewelwell commented on api/tests/unit/segments/test_unit_segments_models.py:
```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
+        feature=feature,
+    )
+
+    sub_rule = SegmentRule.objects.create(type=SegmentRule.ALL_RULE)
```
I'm a little unsure what's going on in this test. This `sub_rule` is not valid since it doesn't have a parent segment or rule and `sub_rule` to me indicates that it's a rule underneath something, whereas below we're using it as a parent of what we define (via the comment) as a parent rule.

zachaysan replied:
Yeah I had it backwards in my head when I was writing it. I've fixed this locally. Either way the test is of a misbehaving rule, but I've added the segment to the rule before also adding the rule which has two parents (needed for the test).

matthewelwell commented on api/tests/unit/segments/test_unit_segments_views.py:
```diff
@@ -175,11 +176,14 @@ def test_audit_log_created_when_segment_updated(project, segment, client):
     assert res.status_code == status.HTTP_200_OK
+
+    # Three audit log records created, two for the primary
+    # segment and one for the newly versioned segment.
```
Hmm... this feels odd. I don't think we should be creating new audit log records for older versions of segments - the operation of updating a segment should only create a single audit log record, the fact that a versioned segment is created is a byproduct of that operation. 

We can override the `get_skip_audit_log` method on the `Segment` (and maybe `SegmentCondition`) models to ensure we're not writing unnecessary audit log records.

zachaysan replied:
Ok this is now done.

matthewelwell commented on api/tests/unit/segments/test_unit_segments_views.py:
```diff
@@ -175,11 +176,14 @@ def test_audit_log_created_when_segment_updated(project, segment, client):
     assert res.status_code == status.HTTP_200_OK
+
+    # Three audit log records created, two for the primary
+    # segment and one for the newly versioned segment.
```
Hmm... this feels odd. I don't think we should be creating new audit log records for older versions of segments - the operation of updating a segment should only create a single audit log record, the fact that a versioned segment is created is a byproduct of that operation. 

We can override the `get_skip_audit_log` method on the `Segment` (and maybe `SegmentRule` / `Condition`?) model to ensure we're not writing unnecessary audit log records.

zachaysan replied:
Ok this is now done.

matthewelwell reviewed (changes requested):
I've added a few more comments but I also noticed that none of the `view` tests actually verify that the clone is created (perhaps because they rely on the number of audit records instead, which I'm suggesting that we remove).

zachaysan reviewed (commented):
One thing I found while I was programming this was that `SegmentRule` is not presently being saved to the audit log. I made methods for skipping the audit log creation in advance of anyone adding audit logging to segment rules, since they will be cloned many times as well. 

Should we make a ticket to add audit logging, including tests for different usage patters triggering audit log, or is it too low of a priority?

zachaysan commented:
> I've added a few more comments but I also noticed that none of the `view` tests actually verify that the clone is created (perhaps because they rely on the number of audit records instead, which I'm suggesting that we remove).

I've added view tests with segment checking including exception handling now.

zachaysan commented on api/segments/models.py:
```diff
@@ -45,14 +57,44 @@ class Segment(
         return "Segment - %s" % self.name
 
+    def get_skip_create_audit_log(self) -> bool:
+        skip = segment_audit_log_helper.should_skip_audit_log(self.id)
```
After trying multiple different approaches including instance variables, `.initial_state`, `.state` all of which were wiped during the save process I finally resorted to using a helper to track the state of which audit logs can be created.

matthewelwell commented on api/segments/models.py:
```diff
@@ -84,6 +126,70 @@ def id_exists_in_rules_data(rules_data: typing.List[dict]) -> bool:
+        self.save()
+        segment_audit_log_helper.unset_skip_audit_log(self.id)
+
+    def deep_delete(self, hard: bool) -> None:
```
I'm still unsure why this is needed vs just the cascade delete. I think the soft delete package also offers a hard delete method as well?

zachaysan replied:
As we discussed during standup, there are issues when running the code in hard delete mode. Waiting on you to take a look.

zachaysan commented on api/segments/models.py:
```diff
@@ -45,14 +57,44 @@ class Segment(
+            return skip
+
+        try:
+            if self.version_of and self.version_of != self:
+                return True
```
Nice catch. Updated here and in the other models.

zachaysan commented on api/segments/models.py:
```diff
@@ -84,6 +126,70 @@ def id_exists_in_rules_data(rules_data: typing.List[dict]) -> bool:
+        cloned_segment.version_of = self
+        cloned_segment.save()
+
+        segment_audit_log_helper.set_skip_audit_log(self.id)
```
We don't want to run the save hook every time with these changes so I think they are better in-place.
